REGRESSION(298556@main): [GTK] Rendering broken in Boxes with 3D passthrough disabled

carlosgcampos opened the pull request:
#### 865be85738f3451d336e0c0b3d4b750c14305f42
<pre>
REGRESSION(298556@main): [GTK] Rendering broken in Boxes with 3D passthrough disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=298776">https://bugs.webkit.org/show_bug.cgi?id=298776</a>

Reviewed by Nikolas Zimmermann.

Do not fallback to DRM API to get main DRM device information when EGL
extension is present but reports nullptr device, which is the case when
software rasterization is used.

Canonical link: <a href="https://commits.webkit.org/300021@main">https://commits.webkit.org/300021@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/190d06c7c629ad468ab0f61e7dcd19687e0f368f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121068 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40764 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127487 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73149 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41466 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49343 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91960 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61176 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124020 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33110 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108526 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72644 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32134 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71078 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/102616 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26808 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130340 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47995 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36474 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100568 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48363 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104693 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100470 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45874 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23926 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44683 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19205 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47853 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53566 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47324 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50671 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49008 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/UIProcess/glib/DRMMainDevice.cpp:
```diff
@@ -189,19 +199,42 @@ const WebCore::DRMDevice& drmMainDevice()
             mainDevice->primaryNode = drmPrimaryNodeDeviceForRenderNodeDevice(mainDevice->renderNode);
+            return;
+        }
+
```
I think some comments when which section is used would be really useful here - I see myself wondering about this in a year 🤣

carlosgcampos replied:
Yes, it's a bit messy.

nikolaszimmermann commented on Source/WebKit/UIProcess/glib/DRMMainDevice.cpp:
```diff
@@ -189,19 +199,42 @@ const WebCore::DRMDevice& drmMainDevice()
+        if (eglDisplay == EGL_NO_DISPLAY)
+            return;
+
+        if (EGLDeviceEXT device = eglDisplayDevice(eglDisplay)) {
```
So this will be nullptr if software rendering is used?

carlosgcampos replied:
No, in case of swrast we get a device and claims to support the extension to get the render node, but it returns nullptr. So, what this patch does is bringing back the previous behavior, so that when de device reports a nullptr render node, we don't try to fallback to DRM API.

nikolaszimmermann commented on Source/WebKit/UIProcess/glib/DRMMainDevice.cpp:
```diff
@@ -189,19 +199,42 @@ const WebCore::DRMDevice& drmMainDevice()
+        if (eglDisplay == EGL_NO_DISPLAY)
+            return;
+
+        if (EGLDeviceEXT device = eglDisplayDevice(eglDisplay)) {
```
auto?

nikolaszimmermann commented on Source/WebKit/UIProcess/glib/DRMMainDevice.cpp:
```diff
@@ -189,19 +199,42 @@ const WebCore::DRMDevice& drmMainDevice()
+        if (EGLDeviceEXT device = eglDisplayDevice(eglDisplay)) {
+            if (auto renderNode = drmRenderNodeDevice(device)) {
+                mainDevice->renderNode = WTFMove(*renderNode);
+                if (!mainDevice->renderNode.isNull()) {
```
If renderNode is null, we don't want the drmFirstDeviceWithRenderNode fallback? When can this happen?

carlosgcampos replied:
No, we only want to use DRM api when the device extension is not supported, but if the device reports nullptr render node we should use that. This broke LIBGL_ALWAYS_SOFTWARE env var, for example. When the env var is passed in a system that supports GBM we ended up using GBM, instead of surfaceless platform and shared memory buffers as expected.

nikolaszimmermann replied:
Thanks for clarifying

nikolaszimmermann reviewed (commented):
I guess we need to verify that the GNOME Boxes problem is gone and imx* remains working first?

nikolaszimmermann reviewed (approved):
I still think some comments would be nice explaining the logic why which fallbacks are used - otherwise needs a lot of domain knowledge to understand it ☺️

carlosgcampos committed:
REGRESSION(298556@main): [GTK] Rendering broken in Boxes with 3D passthrough disabled
https://bugs.webkit.org/show_bug.cgi?id=298776

Reviewed by Nikolas Zimmermann.

Do not fallback to DRM API to get main DRM device information when EGL
extension is present but reports nullptr device, which is the case when
software rasterization is used.

Canonical link: https://commits.webkit.org/300021@main
